src/proj_equirect.jl

```diff
@@ -242,41 +231,109 @@ end
+# works
+Base.parent(M::BlockDiagEquiRect) = M.blocks # for convienience
+Base.length(M::BlockDiagEquiRect) = size(parent(M),3)
+Base.eltype(::Type{BlockDiagEquiRect{B,P,T}}) where {B,P,T} = T 
+Base.firstindex(M::BlockDiagEquiRect) = 1
+Base.lastindex(M::BlockDiagEquiRect) = length(M)
+Base.iterate(M::BlockDiagEquiRect) = (Σ=parent(M) ; isempty(Σ) ? nothing : (Σ[:,:,1],1))
+Base.iterate(M::BlockDiagEquiRect, st) = st+1 > length(M) ? nothing : (parent(M)[:,:,st+1],  st+1)
+
+function Base.getindex(M::BlockDiagEquiRect, i::Int) 
+    1 <= i <= length(M) || throw(BoundsError(M, i))
+    return parent(M)[:,:,i]
 end
 
-function *(B::BlockDiagEquiRect{QUAzFourier}, f::EquiRectQUAzFourier)
-    # TODO: implement S2 multiplication
-    error("not implemented")
+function Base.setindex!(M::BlockDiagEquiRect, m::Matrix, i::Int)
+    1 <= i <= length(M) || throw(BoundsError(M, i))  
+    setindex!(parent(M)[:,:,i], m)
 end
 
-function adapt_structure(storage, L::BlockDiagEquiRect{B}) where {B}
-    BlockDiagEquiRect{B}(adapt(storage, L.blocks), adapt(storage, L.blocks_sqrt), adapt(storage, L.proj))
+# ## Mult (this one is different than M * f since )
+
+*(M::BlockDiagEquiRect{B}, f::EquiRectField) where {B<:Basis} = M * B(f)
+
+function *(M::BlockDiagEquiRect{AzFourier}, f::EquiRectAzFourier)
+    promote_metadata_strict(M.proj, f.proj) # ensure same projection
+    EquiRectAzFourier(@tullio(Bf[p,iₘ] := M.blocks[p,q,iₘ] * f.arr[q,iₘ]), f.metadata)
 end
 
-function simulate(rng::AbstractRNG, L::BlockDiagEquiRect{AzFourier,ProjEquiRect{T}}) where {T}
-    @unpack Ny, Nx, θspan = L.proj
-    z = EquiRectMap(randn(rng, T, Ny, Nx) .* sqrt.(sin.(range(θspan..., length=Ny))), L.proj)
-    sqrt(L) * z
+function *(M::BlockDiagEquiRect{QUAzFourier}, f::EquiRectQUAzFourier)
+    promote_metadata_strict(M.proj, f.proj) # ensure same projection
+    EquiRectQUAzFourier(@tullio(Bf[p,iₘ] := M.blocks[p,q,iₘ] * f.arr[q,iₘ]), f.metadata)
 end
 
+# ## map, e.g. map((Bℓ,fℓ)->sqrt(Bℓ)*fℓ, BlkDiag, field) -> field
+
+function Base.map(fun::Function, M::BlockDiagEquiRect{AzFourier}, f::F) where {F<:EquiRectAzFourier}
+    promote_metadata_strict(M.proj, f.proj) # ensure same projection
+    Mfarr = similar(f.arr)
+    for (Mfc, Mc, fc) in zip(eachcol(Mfarr), M, eachcol(f.arr))
+        Mfc = fun(Mc, fc)
+    end
+    EquiRectAzFourier(Mfarr, f.metadata)
+end 
+
+
+function Base.map(fun::Function, M::BlockDiagEquiRect{QUAzFourier}, f::F) where {F<:EquiRectQUAzFourier}
+    promote_metadata_strict(M.proj, f.proj) # ensure same projection
+    Mfarr = similar(f.arr)
+    for (Mfc, Mc, fc) in zip(eachcol(Mfarr), M, eachcol(f.arr))
+        Mfc = fun(Mc, fc)
+    end
+    EquiRectQUAzFourier(Mfarr, f.metadata)
+end 
+
+# ## I don't understand this yet ??
+# function adapt_structure(storage, L::BlockDiagEquiRect{B}) where {B}
+#     BlockDiagEquiRect{B}(adapt(storage, L.blocks), adapt(storage, L.blocks_sqrt), adapt(storage, L.proj))
+# end
```

Comment:
This is how you make `cu(M::BlockDiagEquiRect)` work to move the object to GPU. This just says recursively move the fields of `BlockDiagEquiRect` to GPU.

Reply:
Ok. I basically turned it off to see what effect it had and as a way to punt the issue of me dropping the `sqrt` field in `BlockDiagEquiRect` (and not sure how to accommodate that change and/or why the `sqrt` field was put in `BlockDiagEquiRect` in the first place)

Reply:
(answered above about `sqrt`)